File: starfightersquadronsweb/src/test/js/AttackDiceTest.js
define(["AttackDice"], function(AttackDice)
{
    "use strict";
    QUnit.module("AttackDice");

    QUnit.test("AttackDice properties", function(assert)
    {
        var dice = new AttackDice(3);
        assert.ok(dice.value(0));
        assert.ok(dice.value(1));
        assert.ok(dice.value(2));
    });

    QUnit.test("addDie()", function(assert)
    {
        // Setup.
        var dice = new AttackDice(1);

        // Run.
        dice.addDie();

        // Verify.
        assert.equal(dice.size(), 2);
        assert.ok(AttackDice.Value.values().vizziniContains(dice.value(1)));
    });

    QUnit.test("blankCount()", function(assert)
    {
        var dice = new AttackDice(1);
        LOGGER.trace("dice = " + dice);
        if (dice.value(0) == AttackDice.Value.BLANK)
        {
            assert.equal(dice.blankCount(), 1);
        }
        else
        {
            assert.equal(dice.blankCount(), 0);
        }
    });

    QUnit.test("criticalHitCount()", function(assert)
    {
        var dice = new AttackDice(1);
        LOGGER.trace("dice = " + dice);
        if (dice.value(0) == AttackDice.Value.CRITICAL_HIT)
        {
            assert.equal(dice.criticalHitCount(), 1);
        }
        else
        {
            assert.equal(dice.criticalHitCount(), 0);
        }
    });

    QUnit.test("focusCount()", function(assert)
    {
        var dice = new AttackDice(1);
        LOGGER.trace("dice = " + dice);
        if (dice.value(0) == AttackDice.Value.FOCUS)
        {
            assert.equal(dice.focusCount(), 1);
        }
        else
        {
            assert.equal(dice.focusCount(), 0);
        }
    });

    QUnit.test("hitCount()", function(assert)
    {
        var dice = new AttackDice(1);
        LOGGER.trace("dice = " + dice);
        if (dice.value(0) == AttackDice.Value.HIT)
        {
            assert.equal(dice.hitCount(), 1);
        }
        else
        {
            assert.equal(dice.hitCount(), 0);
        }
    });

    QUnit.test("newInstance()", function(assert)
    {
        // Setup.
        var dice = new AttackDice(5);

        // Run.
        var result = dice.newInstance();

        // Verify.
        assert.ok(result);
        assert.ok(result !== dice);
        assert.equal(result.size(), dice.size());

        for (var i = 0; i < result.size(); i++)
        {
            assert.equal(result.value(i), dice.value(i));
        }
    });

    QUnit.test("rerollAllFocus()", function(assert)
    {
        var dice;
        do {
            dice = new AttackDice(1);
        }
        while (dice.focusCount() === 0);

        // Run.
        dice.rerollAllFocus();

        // Verify.
        if (dice.value(0) === AttackDice.Value.FOCUS)
        {
            assert.equal(dice.focusCount(), 1);
        }
        else
        {
            assert.equal(dice.focusCount(), 0);
        }
    });

    QUnit.test("rerollBlank() 1", function(assert)
    {
        var dice;
        do {
            dice = new AttackDice(1);
        }
        while (dice.blankCount() === 0);

        // Run.
        dice.rerollBlank();

        // Verify.
        if (dice.value(0) === AttackDice.Value.BLANK)
        {
            assert.equal(dice.blankCount(), 1);
        }
        else
        {
            assert.equal(dice.blankCount(), 0);
        }
    });

    QUnit.test("rerollBlank() 2", function(assert)
    {
        var dice;
        do {
            dice = new AttackDice(2);
        }
        while (dice.blankCount() < 2);

        // Run.
        dice.rerollBlank(2);

        // Verify.
        if (dice.value(0) === AttackDice.Value.BLANK && dice.value(1) === AttackDice.Value.BLANK)
        {
            assert.equal(dice.blankCount(), 2);
        }
        else if (dice.value(0) === AttackDice.Value.BLANK || dice.value(1) === AttackDice.Value.BLANK)
        {
            assert.equal(dice.blankCount(), 1);
        }
        else
        {
            assert.equal(dice.blankCount(), 0);
        }
    });

    QUnit.test("rerollBlankAndFocus() 2, 0", function(assert)
    {
        var dice;
        do {
            dice = new AttackDice(2);
        }
        while (dice.blankCount() < 2);

        // Run.
        dice.rerollBlankAndFocus(2);

        // Verify.
        if (dice.value(0) === AttackDice.Value.BLANK && dice.value(1) === AttackDice.Value.BLANK)
        {
            assert.equal(dice.blankCount(), 2);
        }
        else if (dice.value(0) === AttackDice.Value.BLANK || dice.value(1) === AttackDice.Value.BLANK)
        {
            assert.equal(dice.blankCount(), 1);
        }
        else
        {
            assert.equal(dice.blankCount(), 0);
        }
    });

    QUnit.test("rerollBlankAndFocus() 0, 2", function(assert)
    {
        var dice;
        do {
            dice = new AttackDice(2);
        }
        while (dice.focusCount() < 2);

        // Run.
        dice.rerollBlankAndFocus(2);

        // Verify.
        if (dice.value(0) === AttackDice.Value.FOCUS && dice.value(1) === AttackDice.Value.FOCUS)
        {
            assert.equal(dice.focusCount(), 2);
        }
        else if (dice.value(0) === AttackDice.Value.FOCUS || dice.value(1) === AttackDice.Value.FOCUS)
        {
            assert.equal(dice.focusCount(), 1);
        }
        else
        {
            assert.equal(dice.focusCount(), 0);
        }
    });

    QUnit.test("rerollFocus() 1", function(assert)
    {
        var dice;
        do {
            dice = new AttackDice(1);
        }
        while (dice.focusCount() === 0);

        // Run.
        dice.rerollFocus();

        // Verify.
        if (dice.value(0) === AttackDice.Value.FOCUS)
        {
            assert.equal(dice.focusCount(), 1);
        }
        else
        {
            assert.equal(dice.focusCount(), 0);
        }
    });

    QUnit.test("rerollFocus() 2", function(assert)
    {
        var dice;
        do {
            dice = new AttackDice(2);
        }
        while (dice.focusCount() < 2);

        // Run.
        dice.rerollFocus(2);

        // Verify.
        if (dice.value(0) === AttackDice.Value.FOCUS && dice.value(1) === AttackDice.Value.FOCUS)
        {
            assert.equal(dice.focusCount(), 2);
        }
        else if (dice.value(0) === AttackDice.Value.FOCUS || dice.value(1) === AttackDice.Value.FOCUS)
        {
            assert.equal(dice.focusCount(), 1);
        }
        else
        {
            assert.equal(dice.focusCount(), 0);
        }
    });
});

File: starfightersquadrons/src/test/java/org/vizzini/starfightersquadrons/AttackDiceTest.java
package org.vizzini.starfightersquadrons;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.fail;

import org.junit.Test;
import org.vizzini.starfightersquadrons.AttackDice;
import org.vizzini.starfightersquadrons.AttackDice.Value;

/**
 * Provides tests for the <code>AttackDice</code> class.
 */
public final class AttackDiceTest
{
    /**
     * Test the <code>changeAllToValue()</code> method.
     */
    @Test
    public void changeAllToValue()
    {
        // Setup.
        final int size = 3;
        final AttackDice dice = new AttackDice(size);
        final int focusCount = dice.getFocusCount();
        final int criticalHitCount = dice.getCriticalHitCount();

        // Run.
        dice.changeAllToValue(Value.FOCUS, Value.CRITICAL_HIT);

        // Run / Verify.
        assertThat(dice.getFocusCount(), is(0));
        assertThat(dice.getCriticalHitCount(), is(focusCount + criticalHitCount));
    }

    /**
     * Test the <code>changeFirstToValue()</code> method.
     */
    @Test
    public void changeFirstToValue()
    {
        // Setup.
        final int size = 3;
        final AttackDice dice = new AttackDice(size);
        final int focusCount = dice.getFocusCount();
        final int criticalHitCount = dice.getCriticalHitCount();

        // Run.
        dice.changeFirstToValue(Value.FOCUS, Value.CRITICAL_HIT);

        // Run / Verify.
        assertThat(dice.getFocusCount(), is(Math.max(focusCount - 1, 0)));
        assertThat(dice.getCriticalHitCount(), is(focusCount == 0 ? criticalHitCount : criticalHitCount + 1));
    }

    /**
     * Test the <code>rollAdditionalDie()</code> method.
     */
    @Test
    public void rollAdditionalDie()
    {
        // Setup.
        final int size = 3;
        final AttackDice dice = new AttackDice(size);

        // Run.
        dice.rollAdditionalDie();

        // Run / Verify.
        assertThat(dice.size(), is(size + 1));
    }

    /**
     * Test the <code>spendFocusToken()</code> method.
     */
    @Test
    public void spendFocusToken()
    {
        // Setup.
        final int size = 3;
        final AttackDice dice = new AttackDice(size);
        final int hitCount = dice.getHitCount();
        final int criticalCount = dice.getCriticalHitCount();
        final int focusCount = dice.getFocusCount();
        final int blankCount = dice.getBlankCount();

        // Run.
        dice.spendFocusToken();

        // Run / Verify.
        assertThat(dice.getHitCount(), is(focusCount + hitCount));
        assertThat(dice.getCriticalHitCount(), is(criticalCount));
        assertThat(dice.getFocusCount(), is(0));
        assertThat(dice.getBlankCount(), is(blankCount));

        for (int i = 0; i < size; i++)
        {
            assertNotNull(dice.getValue(i));
        }
    }

    /**
     * Test the <code>AttackDice()</code> method.
     */
    @Test
    public void testConstructorBadSize()
    {
        try
        {
            new AttackDice(-1);
            fail("Should have thrown an exception");
        }
        catch (final IllegalArgumentException e)
        {
            assertThat(e.getMessage(), is("size is less than zero."));
        }

        try
        {
            new AttackDice(-2);
            fail("Should have thrown an exception");
        }
        catch (final IllegalArgumentException e)
        {
            assertThat(e.getMessage(), is("size is less than zero."));
        }
    }

    /**
     * Test the <code>values()</code> method.
     */
    @Test
    public void values()
    {
        // Setup.
        final int size = 3;
        final AttackDice dice = new AttackDice(size);

        // Run / Verify.
        for (int i = 0; i < size; i++)
        {
            assertNotNull(dice.getValue(i));
        }
    }
}
